src/content/post/20161122_drupal_application_bottlenecks_avoid_with_profiling.md

```diff
@@ -0,0 +1,152 @@
++++
+
+### TL;DR
+
+This article is focused on the possible *bottlenecks of a web application*, how you can identify it to improve the performance of your code.
+
+We talk about profiling with *Blackfire*, PHP web app focused on *Drupal*. LAMP is the reference stack.
+
+<!-- more -->
+
+Application's performance is one of characteristics what separates a job from a good job, and demonstrates quality work.
+
+A customer's call about website performances, if server infrastructure is adequate to requests, most probably indicates some errors on our code.
+
+Google can be our friend if we find some 'web application performance checklist' but we have a problem, now. We need to identify and resolve our performance issues.
+
+A little disclaimer: this article is focused on a Drupal application, but information which follow are useful (I hope!) regardless of framework that you have used.
+
+### Application's check
+
+1. Check caches
+    1. Check Page cache, block cache and view cache
+2. Aggregate and minify CSS and js
+3. Deactivate development component
+    1. Modules like _devel_, _update_status_, _performance_ are really performances eaters.
+4. Deactivates _cron_ automation
+    1. Run tasks autonomously by night or in the hours with light traffic if it's possible. Remember to demand critical tasks or heavy tasks to external workers.
+5. Deactivates database logger
+    1. Move logs to _syslog_
+6. Check images size
+    1. Use opportune _image_styles_
+7. Check retrieval process of external resources (feed, streams, etc.)
+
+### Detection of application's bottleneck
+
+Once make previous checks, probably now our server take a sigh of relief, but _most_ probably this is not enough: we need with precision (specially if we don't have precise reference to slow pages) because we have an high server load. We need to collect data to use to apply right improvement, yes, I'm talking about [profiling](https://en.wikipedia.org/wiki/Profiling_(computer_programming)).
+
+#### Application side
+
+On Drupal 7, we can use the [Performance Logging](https://www.drupal.org/project/performance) module, to profile time spent to generate pages, memory consumption, and how many queries we have made. We need only to navigate our site to populate the data table. On Drupal 8 there is the very good module called _Web Profiler_ joined in the [Devel project](https://www.drupal.org/project/devel).
+
+#### Server side
+
+Servers, but generally the hardware which our application runs on, with right tools become our better friend to help us to find reasons of application's bottlenecks. There is a lot of profiler, but an excellent choice is [Blackfire](http://blackfire.io/): free, flexible and very light (his agent can be installed on the production servers with a really really thin overhead). The feature which makes me love Blackfire are his profiler result: a simple and clear diagram of entire code flow from the request response to the return of the last function called.
```

Comment:
Given the point above, I'd wait you make some clear, then help you fixing some grammar mistake I've found.
Thanks